src/base/bittorrent/sessionimpl.cpp

```diff
@@ -2711,8 +2711,39 @@ bool SessionImpl::addTorrent_impl(const TorrentDescriptor &source, const AddTorr
     if (m_loadingTorrents.contains(id) || (infoHash.isHybrid() && m_loadingTorrents.contains(altID)))
         return false;
 
-    if (findTorrent(infoHash))
+    if (Torrent *torrent = findTorrent(infoHash))
+    {
+        // a duplicate torrent is being added
+
+        if (hasMetadata)
+        {
+            // Trying to set metadata to existing torrent in case if it has none
+            torrent->setMetadata(*source.info());
+        }
+
+        if (!isMergeTrackersEnabled())
```

Comment:
Shouldn't this be checked before `hasMetadata`? The operation above  is wasted if this is false.